First chunk of cell type teaching text (#595)
AlexsLemonade/training-modules: scRNA-seq-advanced/02-celltype_assignment.Rmd

Much of the material in this notebook is directly inspired by, and draws heavily on, material presented in the book [_Orchestrating Single Cell Analysis with Bioconductor_](http://bioconductor.org/books/3.16/OSCA/). 

The data we will use for this notebook is derived from a [10x Genomics dataset of human peripheral blood mononuclear cells (PBMCs)](https://software.10xgenomics.com/single-cell-gene-expression/datasets/6.0.0/10k_PBMCs_TotalSeq_B_3p.)
These data include both single cell RNA-seq counts and quantification of antibody-derived tags (ADTs) performed by sequencing short DNA barcodes attached to specific antibodies. 
This type of ADT sequencing with single cells is commonly known as CITE-seq, after the protocol developed by [Stoeckius _et al._ (2017)](https://doi.org/10.1038/nmeth.4380).  
The antibodies used here are the [The TotalSeq™-B Human TBNK Cocktail](https://www.biolegend.com/en-us/products/totalseq-b-human-tbnk-cocktail-19043), a set of antibodies designed to react with immune cell surface markers.

> The data here have already been filtered, normalized, and had dimension reductions calculated for the scRNA-seq data.

Reviewer: ```suggestion
The data here have already been filtered, normalized, and had dimension reductions calculated for the single-cell RNA-seq data.
```
This is nit-picky but in the sentence above you used `single-cell RNA-seq` so I would just stick with one throughout the notebook.

The ADT data has also been separately filtered and normalized.

> For details about how to these tasks were performed with data that has been processed with Cell Ranger, have a look at our `exercise_01` notebook in this directory.

Reviewer: ```suggestion
For details about how to perform these tasks with data that has been processed with Cell Ranger, have a look at our `exercise_01` notebook in this directory.
```

> For details about how to perform these tasks with data that has been processed with Cell Ranger, have a look at our `exercise_01` notebook in this directory.

Reviewer 2: Noting this notebook links the relevant OSCA chapter too so that's set!
Eventually fill the nb name all the way in, once the name is definitely set.

### Input and output

As mentioned, our input file here is a single normalized and processed SCE object, stored as an `rds` file. 
That should be all we need to read in!

Our output will be a table of per-cell information, which will include the cell type assignments we have made throughout this notebook.

> We aren't planning any significant modifications of the underlying data, so we won't bother re-saving it as a new object this time.

Reviewer: I wonder if we actually do want to show saving the cell types as a column in the colData of the SCE object and then saving the modified SCE object?

> We aren't planning any significant modifications of the underlying data, so we won't bother re-saving it as a new object this time.

Author: I do show adding them in as a column (for plotting), but I didn't bother saving the full object, as that seemed a bit unneccessary for the reasons stated. But I wanted to save something...

```{r filepaths, live=TRUE}
# directory for the input data
data_dir <- file.path("data", "PBMC-TotalSeqB", "normalized")

# the input file itself
sce_file <- file.path(data_dir, "PBMC_TotalSeqB_normalized_sce.rds")

# A directory to store outputs
analysis_dir <- file.path("analysis", "PBMC-TotalSeqB")

# check whether the output directory exists 
if (!dir.exists(analysis_dir)){
  dir.create(analysis_dir, recursive = TRUE)
}

# output table path
cellinfo_file <- file.path(analysis_dir, "PBMC_TotalSeqB_cellinfo.tsv")
```

## Exploring a CITE-seq `SingleCellExperiment`

Now that the preliminary setup is out of the way, we can get started. 
First we will read in the `SingleCellExperiment` from the input file we defined earlier.

```{r read SCE, live=TRUE}
# read in the SCE file

> sce <- readRDS(sce_file)

Author: This is definitely a debate I have in my mind every time... same with `saveRDS` vs `write_rds`...

# print a summary of the SCE
sce
```
This should look similar to the SCE objects that we have seen before, containing `counts` and `logcounts` assays where each cell is a column and each row is a gene.
We also have some of the `rowData`, `colData` and reduced dimension matrices that we have seen before.

But where are the data from the ADTs?
We wouldn't necessarily want those stored in the main data matrices, as the characteristics of ADT barcode data is going to be quite different from gene expression data.

> To keep the ADT data separate, we have split this data off into an _alternative experiment_ (`altExp`) slot.

Author: I read this in exactly the opposite way! My phrase reads to me as something that happened in the past, yours as something that we have yet to do. (I like the "from RNA" part though)

You can see the name of this `altExp` on the line `altExpNames` above. 
We _could_ have more than one type of alternative experiment, but in this case, just the one.

To access the contents of the `altExp` slot, we can use the `altExp()` function.
Let's look at what we have in that slot:

```{r view altExp, live=TRUE}

> altExp(sce, "ADT")

Reviewer: I think I would just add a comment above this line that you are printing out the alternative experiment by providing the name of the experiment

```

It is another `SingleCellExperiment`! 
Inception!
Let's look at that embedded SCE more closely.

The first thing to note is that we have the same number of columns for this `altExp` as we did with the full SCE. 
Those corresponded to the individual cells before, and still do!

There are only 10 rows, however, and these now correspond to the ADTs that were assayed by this particular experiment. 

> Just as we did with the full SCE, we can use `rowData()` to look at table of metadata for each of these rows.

Reviewer: ```suggestion
Just as we did with the full SCE, we can use `rowData()` to view the table containing metadata associated with each of these rows.
```
I found this slightly confusing, so some suggested rewording to make it clear that there is only one table.

We'll add the `altExp()` function to point it to the embedded object we are interested in. 
Since there is only one `altExp`, we don't need the second argument (`"ADT"`) that we used above; the default behavior of `altExp()` is to just give us the first `altExp`, and that is all we need.

```{r adt rows, live=TRUE}
# What proteins were assayed?
adt_tags <- rowData(altExp(sce))
adt_tags
```

You can see here that the names and symbols of the tags used, along with the designation that all have an "Antibody Capture" type.
One you might note looks different is the `IgG1` control, which is actually a mouse antibody used as a negative control. 

### Clustering redux

While these data did have dimension reduction performed, they were not analyzed for clustering yet.

Let's assign some clusters to our cells, using graph-based clustering and default parameters, taking as input the PCA matrix that was previously calculated.
Note that this PCA matrix and the UMAP built from it were derived from the gene expression data, so the clustering is going to reflect the gene expression data only.

> While we have the ADT data, it is _not_ being used for this stage of the analysis.

Reviewer 2: > `mainExp()` doesn't exist

Ah yes I guess I meant this "conceptually"...this is how I keep referring to it in my head but it's not really real!

```{r}
# perform clustering
nn_clusters <- bluster::clusterRows(
  # PCA input
  reducedDim(sce, "PCA"), 
  # graph clustering & parameters
  bluster::NNGraphParam()
)

# add clusters to colData
sce$nn_cluster <- nn_clusters
```

Now we can plot the clusters we have identified with our old friend `scater::plotUMAP()`.

```{r plot clusters}
# plot clusters
scater::plotUMAP(sce, colour_by = "nn_cluster") + 
  # rename the legend

>   guides(colour = guide_legend(title = "Cluster"))

Author: Sadly, no.

```
But what are these clusters, really? 
Do they correspond to particular cell types that we are interested in?

Does it bother you that we just used the default nearest-neighbor graph clustering parameters?

> Do you know what those were?

Reviewer: I love this.

## Investigating cell types

### Using cell surface marker data

> The first way we will work on to identify the cell types of the individual cells in our sample is to work with the ADT markers that are part of this data set. 

Reviewer 2: Tried to reduce the prepositional phrases here. I think it's implicit that the cells are in our sample! but we do want to keep "individual cells" as a contrast to cell typing clusters/metacells (even if those aren't here in the end).

```suggestion
The first way we will identify cell types of individual cells is to use the ADT normalized counts.
```

> These antibody markers were (hopefully) chosen for their relevance to the cell population we are working with.

Reviewer 2: ```suggestion
These antibody markers were (hopefully) chosen for their relevance to the sequenced cell population.
```

> The first marker we will look at is `CD3`, which is a protein complex that is found on the surface of T cells.

Author: I can't decide. I don't feel like it is really needed, and it is pretty easily searchable. But it isn't terrible to include.

> The first marker we will look at is `CD3`, which is a protein complex that is found on the surface of T cells.

Reviewer 2: I certainly don't think it's _needed_. I think anyone who is maybe less familiar would just google on their own anyways as you say.

> We can plot where this is found using the same `plotUMAP()` function as before. 

Reviewer 2: We're not plotting  _presence_ ("where this is found") of CD3 but rather thwe _amount of_ CD3. 

suggested rephrase - 
```suggestion
We can again use the `plotUMAP()` function to color cells by `CD3` expression levels. 
```

> Note that the function is able to look for plotting data in fields found in the `colData` object (as we used it above), in the gene expression matrix (as we used it in the previous notebook), *AND* in `altExp` matrices!

Reviewer 2: I again think we want to be explicit about main vs alt with some of that terminology, and also by now I forgot we plotted clusters.

```suggestion
Note that this function can plot data from the `colData` table (as we used it above when plotting clusters), in the `mainExp` gene expression matrix (as we used it in the previous notebook), *AND* in `altExp` tables and matrices!
```

> So all we need to do to color by the ADT counts (as normalized into the `logcounts` matrix) is give the name of the tag that we want to plot in the `colour_by` argument.

Reviewer 2: ```suggestion
So, to color by ADT counts (as normalized into the `logcounts` matrix), all we need to do is provide the tag name that we want to plot in the `colour_by` argument.
```

```{r plot CD3, live = TRUE}
# plot CD3 expression
scater::plotUMAP(sce, colour_by = "CD3")
```

It appears that we have a number of potential T cells down in the lower left!

Let's look at a couple of other markers to try to break those up more specifically.

> Two other markers of relevance to the T cells are `CD4` and `CD8`.

Reviewer 2: Again if we want to link a source, even though this is pretty standard.

The `CD4` complex is present in helper T cells (hence their other common name, CD4+ T cells).
By contrast, the `CD8` complex is found on killer T cells (CD8+ cells).

Let's plot the ADT results for those two markers as well below:

```{r plot CD4, live=TRUE}
# plot CD4 marker
scater::plotUMAP(sce, 
                 colour_by = "CD4")
```

```{r plot CD8, live=TRUE}

> # plot CD8 marker

Reviewer 2: Note that after this chunk/before next section, maybe add in some discussion scaffolding about how this approach is much more qualitative and cell typing is hard.

scater::plotUMAP(sce, 
                 colour_by = "CD8")
```

### Rule-based classification

> When we have markers like these, with clear and established cell types, we might want to start our classification by working with these markers alone.

Reviewer 2: This sentence feels confusing since didn't we just do that?

Author: Will revise, but I want to change from plotting levels to actual classification

> Such classification might be considered as analogous to a cell-sorter assay, where we would set up some rules to look at a few markers for each cell and use those to assign a cell type classification.

Reviewer 2: lot of "might"s
```suggestion
Such classification could be considered as analogous to a cell-sorter assay, where we would set up some rules to look at a few markers for each cell and use those to assign a cell type classification.
```

Also, give an example of what "rule" means here to lead into why things like thresholds (next sentence) are relevant at all.

To do this, we will need to make some decisions, like what thresholds should we use to determine whether a cell is or is not expressing a particular marker. 

> In general, if a marker is going to be useful for this type of assignment, we would like to see a bimodal plot for the marker in question that can be used to separate the population into two groups of cells based on the level of that marker.

Reviewer 2: ```suggestion
In general, markers that are useful for this cell-typing approach would have a bimodal distribution of expression levels which can be used to separate the population into two groups of cells.
```
Add a second sentence here about what the bimodal means here (foreground/background).

Reviewer: I would echo the explanation here about first defining what a "rule" is to assign a cell type and then introducing the idea that although a gene could be a marker for a set of cells it's not going to be binary across cells, there is going to be some level of background and how do you define what that that background is.

> To look at the expression levels of each of the marker genes we are interested in, we will pull out the expression values for the markers from the SCE object and do some data wrangling. 

Reviewer 2: Add something like why - this will tells us whether those markers are suitable for cell annotation.

> The data we are interested is the normalized count data for the ADT tags, which is stored in the `logcounts` assay.

Reviewer 2: ```suggestion
We are interested in the normalized counts for the ADT tags, which are stored in the `logcounts` assay of the `altExp`.
```

> If you recall, this matrix is stored with the columns as cells and rows as markers, but we really want it the other way, so we will first transpose the data, then convert it to a data frame for our next steps.

Reviewer 2: > but we really want it the other way

Need more explanation here

> Because the SCE object stores the main data matrices in a specialized format, we have to do one extra step convert it first to a "regular" R matrix or R won't know how to convert it to a data frame.

Reviewer 2: "main" isn't my favorite here given `main` vs `altExp` and we're talking about assays here.
I suggested "optimized for memory" as well to contextualize what this specialized format is. I can't decide if this is actually a "non-technical" term or if it's still too loaded though.

```suggestion
Because the SCE object stores the assays matrices in a specialized format optimized for memory, we have to do one extra step convert it first to a "regular" R matrix or R won't know how to convert it to a data frame.
```

Author: Yeah, I didn't want to get into this, as the real truth is that ADT matrices may or may not be stored as sparse (they are not sparse!), so I think I'm going to stick with the hand waving...

> ```{r extract ADT}

Reviewer 2: We'll probably want to demo this one step at a time while teaching, just noting. No change requested.

Author: I was actually thinking probably not... I didn't have this as a live chunk. I want people to stay mostly focused on the results at this point, not on the details of the coding.

Reviewer 2: Sure, sounds fine!

# convert logcounts data to a data frame
adt_df <- logcounts(altExp(sce)) %>%
  t() %>% # transpose
  as.matrix() %>% # convert to matrix
  as.data.frame() # convert to data frame

# view the data frame
head(adt_df)
```

> If we just wanted to plot one of these tags, we could do so right away, but with a bit more data wrangling, we can convert these results into a "tidier" format, that will allow us to take full advantage of `tidyverse` tools!

Reviewer 2: There's a little more context here that is needed - we're transitioning away from `plotUMAP` and into directly plotting with `ggplot2`. But why do we have to do that if we have the convenient `plotUMAP` function? Because plot moar!

> Right now the data is in a "wide" format, with each column a different tag. 

Reviewer 2: ```suggestion
Right now the data is in a "wide" format, such that each column is a different tag. 
```

But the data in all of the columns is the same type, and measures something similar: the normalized count of an ADT.
One could even argue that each row contains 10 different observations, where the "tidy" data ideal, as espoused by [Wickham (2014)](https://doi.org/10.18637/jss.v059.i10), requires a single observation per row, a "long" format.

> This long format will have one column that tells us which ADT was measured and a second column with the measurement value itself.

Reviewer 2: Something worth pointing with reshaping (might even be worth writing out explicitly somehow) is that folks newer to tidy formats are often sad at the "repetition" in the resulting long df, e.g. multiple rows for the same cell. We'll want to point out that this is FINE and NOT REDUNDANT and actually can make some tasks EASIER TO DO. 

In this case we don't actually have a cell_id column though so they might not have this explicit confusion in this case.

> We can perform this conversion using the `tidyr::pivot_longer()` function.

Reviewer 2: ```suggestion
We can perform this conversion using the [`tidyr::pivot_longer()`](https://tidyr.tidyverse.org/articles/pivot.html) function.
```

Might note that older versions of `tidyr` used `gather()` (inverse `spread()`) and they might see that in the wild, and package versions matter!

Following conversion, we can filter to just the ADTs that we care about.

> which allows us to convert our data frame with one column per tag into a data frame with separate columns for the tag id (`ADT`) and the expression value (`logcount`).

Reviewer 2: Something's not done here

> ```{r pivot longer, live=TRUE}

Reviewer 2: Let's get an image in here before the chunk. Maybe Fig 12.2 here? https://r4ds.had.co.nz/tidy-data.html

Reviewer 2: Actually, maybe also some before/after object screen shots would be good too.

adt_df_long <- adt_df %>%
  # pivot to long format
  tidyr::pivot_longer(
    everything(), # use all columns

>     names_to = "ADT", # convert row names to a column called "ADT"

Reviewer 2: ```suggestion
    names_to = "ADT", # Convert existing column names to become a new column called "ADT"
```

Author: Line length limits mean we can't do comments this long. The ones I have may be too long already! 

The normal solution would normally be to make these as multi-line comments, but I don't want to do that in this case, because of how `exrcise` will treat the comment-only lines.

>     values_to = "logcount" # name the value column "logcount"

Reviewer 2: I'm too wordy here probably
```suggestion
    values_to = "logcount" # Send the values they were in the "ADT" columns into their column called "logcount"
```

  ) %>%
  # filter to tags we are interested in
  dplyr::filter(ADT %in% c("CD3", "CD4", "CD8"))

# look at the resulting df
head(adt_df_long)
```

Now we can make a density plot with `ggplot2` for all three ADTs we are interested in at once.

```{r plot ADTs, live=TRUE}
# plot logcounts by ADT
ggplot(adt_df_long, aes(x = logcount, fill = ADT)) + 
  geom_density() + # density plot
  facet_wrap(~ADT, ncol = 1) + # facet by ADT

>   theme_bw() + # nicer theme

Reviewer 2: ```suggestion
  theme_bw() + # nicest theme
```
if you're feeling extra spicy

  theme(legend.position = "none") # no legend needed
```

These look pretty good!

> There seem to be different baseline levels for each of the different tags, but each has a pretty nice bimodal distribution.

Reviewer 2: ```suggestion
There seem to be different baseline levels for each of the different tags, but each has a pretty clear bimodal distribution.
```

We can now use the values from these plots to construct a set of rules to classify the T cells. 

> We will do this using the "wide" data frame from earlier. 

Reviewer 2: Someone will ask "why/when use wide vs long" and the answer will just have to be "depends on what you are doing and which format makes more sense for your task."

> The thresholds we are using here were found just "by eye", so this is not a particularly principled method of cell type assignment, but it can be fairly effective.

Reviewer 2: ```suggestion
The thresholds we are using here were identified just "by eye", so this is not a particularly principled method of cell type assignment, but it can be fairly effective.
```

> but it can be fairly effective.

might there be a pub or so showing it can be fairly effective?

Author: I'm not sure I have one, except to say "I've seen this in publications"

Reviewer: I was trying to think of publications that show this and I don't have any on hand (I'm sure there is one that is older that I just can't find). The ones I have seen are all field-specific and dependent on the types of cells being annotated. The other option is to go back to the earlier Seurat papers that use expression of marker genes to define cell types, but I don't really think that you need that here. I think what you have is okay without a publication.

> Here we are assigning only three cell types: cells that do not fit any of these criteria will be set as `NA`.

Reviewer 2: We could also do a `TRUE ~ "Other"` instead of defaulting to NA here, but I don't have a strong preference.

```

### Rule-based classification

Plotting the levels of the ADTs provides a nice visual representation, but what we really want to do is to turn these values into specific cell-type assignments for each cell.
Such classification could be considered as analogous to a cell-sorter assay, where we would set up some rules to look at a few markers for each cell and use those to assign a cell type.
The simplest type of rule might be one where we use a threshold to call a marker as present or absent, then to use the presence of a marker to indicate a specific cell type.

To do this, we will need to make some decisions, such as the thresholds should we use to determine whether a cell is or is not expressing a particular marker. 
In general, markers that are useful for this cell-typing approach will have a bimodal distribution of expression levels which can be used to separate the population into two groups of cells.
One group of cells will have only a background level signal for each marker (due to non-specific binding or other factors), while the other group, those that express the protein, will have a much higher level of binding and higher counts.

> To assess whether the ADTs we have chosen have a useful distribution of expression values, and to identify thresholds we might use, we would like to plot each ADT tag. To do this, we will pull out the expression values for these markers from the SCE object and do some data wrangling. 

Author: ```suggestion
To assess whether the ADTs we have chosen have a useful distribution of expression values, and to identify thresholds we might use, we would like to plot each ADT tag.
To do this, we will pull out the expression values for these markers from the SCE object and do some data wrangling. 
```

To access the contents of the `altExp` slot, we can use the `altExp()` function.
Let's look at what we have in that slot:

```{r view altExp, live=TRUE}
# print a summary of the 'ADT' altExp
altExp(sce, "ADT")
```

It is another `SingleCellExperiment`! 
Inception!
Let's look at that embedded SCE more closely.

The first thing to note is that this `altExp` has the same number of columns as did the main SCE object. 
Those corresponded to the individual cells before, and still do!

There are only 10 rows, however, and these correspond to the ADTs that were assayed by this particular experiment. 
Just as we did with the full SCE, we can use `rowData()` to view the table containing metadata associated with each of these rows.
We'll add the `altExp()` function to point it to the embedded object we are interested in. 
Since there is only one `altExp`, we don't need the second (name) argument (`"ADT"`) that we used above; the default behavior of `altExp()` is to just give us the first `altExp`, and that is the one (and only) that we need.

```{r adt rows, live=TRUE}
# What proteins were assayed?
adt_tags <- rowData(altExp(sce))
adt_tags
```

> You can see here that the names and symbols of the tags used, along with the designation that all have an "Antibody Capture" type (as opposed to "Gene Expression" for the RNA data).

Reviewer: ```suggestion
You can see here the names and symbols of the tags used, along with the designation that all have an "Antibody Capture" type (as opposed to "Gene Expression" for the RNA data).
```
I think this is what you meant.

```{r plot CD3, live = TRUE}
# plot CD3 expression
scater::plotUMAP(sce, colour_by = "CD3")
```

It appears that we have a number of potential T cells down in the lower left!

Let's look at a couple of other markers to try to break those up more specifically.

Two other markers of relevance to the T cells are `CD4` and `CD8`.
The `CD4` complex is present in helper T cells (hence their other common name, CD4+ T cells).
By contrast, the `CD8` complex is found on killer T cells (CD8+ cells).

Let's plot the ADT results for those two markers as well below:

```{r plot CD4, live=TRUE}
# plot CD4 marker
scater::plotUMAP(sce, 
                 colour_by = "CD4")
```

```{r plot CD8, live=TRUE}
# plot CD8 marker
scater::plotUMAP(sce, 
                 colour_by = "CD8")
```

### Rule-based classification

Plotting the levels of the ADTs provides a nice visual representation, but what we really want to do is to turn these values into specific cell-type assignments for each cell.
Such classification could be considered as analogous to a cell-sorter assay, where we would set up some rules to look at a few markers for each cell and use those to assign a cell type.

> The simplest type of rule might be one where we use a threshold to call a marker as present or absent, then to use the presence of a marker to indicate a specific cell type.

Reviewer 2: ```suggestion
The simplest type of rule might be one where we use a threshold to call a marker as present or absent, and then use the presence of a marker to indicate a specific cell type.
```

> To do this, we will need to make some decisions, such as the thresholds should we use to determine whether a cell is or is not expressing a particular marker. 

Reviewer: ```suggestion
To do this, we will need to make some decisions, such as the thresholds we should use to determine whether a cell is or is not expressing a particular marker. 
```

This long format will have one column that tells us which ADT was measured and a second column with the measurement value itself.

We can perform this conversion using the [`tidyr::pivot_longer()`](https://tidyr.tidyverse.org/articles/pivot.html) function,
which allows us to convert our data frame with one column per tag into a data frame with separate columns for the tag id (`ADT`) and the expression value (`logcount`).
Following conversion, we will filter to just the ADTs that we care about.

```{r pivot longer, live=TRUE}
adt_df_long <- adt_df %>%
  # pivot to long format
  tidyr::pivot_longer(
    everything(), # use all columns
    names_to = "ADT", # convert row names to a column called "ADT"
    values_to = "logcount" # name the value column "logcount"
  ) %>%
  # filter to tags we are interested in
  dplyr::filter(ADT %in% c("CD3", "CD4", "CD8"))

# look at the resulting df
head(adt_df_long)
```

Now we can make a density plot with `ggplot2` for all three ADTs we are interested in at once.

```{r plot ADTs, live=TRUE}
# plot logcounts by ADT
ggplot(adt_df_long, aes(x = logcount, fill = ADT)) + 
  geom_density() + # density plot

>   facet_wrap(~ADT, ncol = 1) + # facet by ADT

Reviewer 2: ```suggestion
  facet_wrap(vars(ADT), ncol = 1) + # facet by ADT
```

Reviewer 2: we're headed this direction I think?

Author: We are, but I think I'm going to go all the way to `facet_grid` here, as the explicit rows is nice:
```suggestion
  facet_grid(rows = vars(ADT)) + # facet by ADT
```

  )
```

Now we will want to add the cell types we have assigned back to our original SCE object.

> We can do that by defining a new column name, `threshold_celltype` that will be added to the `colData` object.

Reviewer 2: maybe `threshold_adt_celltype`? since it didn't use RNA (which we later say can also be used for thresholding!).

Creating and assigning values to this column can be done with the `$` shortcut, and then we can plot our results with the `plotUMAP()` function as before.

> ```{r plot thresholds}

Reviewer 2: Should point out that even though this is ADT information we still are looking at our RNA-level UMAP! So, this data indeed goes into the main experiment coldata, not the ADT experiment coldata.

sce$threshold_celltype <- adt_df$celltype
scater::plotUMAP(sce, 
                 colour_by = "threshold_celltype") + 

>   guides(color = guide_legend(title = "Cell type"))

Reviewer 2: again does `labs()` work here?

```

How did we do? 

Note that while we applied this technique to assign cell types using the ADT data, we could use the same type of procedure using gene expression data alone, or a combination of gene expression data and tag data.

However, what we did here was very ad-hoc and quite manual!
We didn't calculate any statistics, and we had to look at every tag we were interested in to pick thresholds.

> A different data set might have different background levels, which would require different thresholds. 

Reviewer 2: This is the first time "background levels" as a term is used. We should introduce it with the bimodal introduction (explicit BG/FG designations; I commented above as well).

Reviewer 2: Hm also, we could explore here "what if we had chosen different thresholds" and how would our results change? We could do something like count the number of cells per cell type (in particular the unclassified!) and see the difference.

Author: This whole section is a "you could, but probably don't," so I don't want to spend too much time on it!

So while this technique is quite limited in its utility.

> It might be good for some simple experiments, but it might not translate well to more complex data sets with multiple samples.

Reviewer 2: one sentence?
